Complete editorial overhaul of usubscription.proto (#323)
eclipse-uprotocol/up-spec: up-l3/usubscription/v3/README.adoc

The remainder of this document defines the use cases, interfaces, state machines and application flows to support cross-system tracking of subscriptions, especially to/from remote systems.

---

[.specitem,oft-sid="dsn~usubscription-service_name~1",oft-needs="impl,utest"]
The uSubscription service name *MUST* be `core.usubscription`.

[.specitem,oft-sid="dsn~usubscription-uentity_id~1",oft-needs="impl,utest"]

> The uSubscription service uEntity Identifier (`ue_id`) *MUST* be 0.

Reviewer: ...uEntity _type_ identifier must be 0x0000.

Author: ok

The remainder of this document defines the use cases, interfaces, state machines and application flows to support cross-system tracking of subscriptions, especially to/from remote systems.

---

[.specitem,oft-sid="dsn~usubscription-service_name~1",oft-needs="impl,utest"]
The uSubscription service name *MUST* be `core.usubscription`.

[.specitem,oft-sid="dsn~usubscription-uentity_id~1",oft-needs="impl,utest"]
The uSubscription service uEntity Identifier (`ue_id`) *MUST* be 0.

The USubscription service API is defined using UML2 notation.

[mermaid]
ifdef::env-github[[source,mermaid]]
----

> classDiagram
> class USubscription {
>     subscribe(topic: UUri, expiry: UInt64[0..1], sample-period: UInt32[0..1]) topic: UURI, status: SubscriptionStatus
>     unsubscribe(topic: UUri)
> }

Reviewer: ```suggestion
class USubscription {
    <<interface>>
    subscribe(topic: UUri, expiry: UInt64 [0..1], sample-period: UInt32 [0..1]) : UUri, SubscriptionStatus
    unsubscribe(topic: UUri)
}

class SubscriptionStatus {
    <<enumeration>>
    unsubscribed
    subscribePending
    subscribed
    unsubscribePending
}

USubscription ..> SubscriptionStatus
```

Author: yes

The remainder of this document defines the use cases, interfaces, state machines and application flows to support cross-system tracking of subscriptions, especially to/from remote systems.

---

[.specitem,oft-sid="dsn~usubscription-service_name~1",oft-needs="impl,utest"]
The uSubscription service name *MUST* be `core.usubscription`.

[.specitem,oft-sid="dsn~usubscription-uentity_id~1",oft-needs="impl,utest"]

> The uSubscription service uEntity Identifier (`ue_id`) *MUST* be 0x0000.

Reviewer: ```suggestion
The uSubscription service uEntity type identifier *MUST* be 0x0000.
```
This is only the lower 16 bits of the 32-bit ue_id ...

Author: ok got it

A _Subscriber_ is a uEntity that is interested in a particular _topic_ and wants to receive messages that have been published to that topic by other uEntities. Subscribers typically use the {communication_layer_api_ref} for this purpose.

[.specitem,oft-sid="dsn~usubscription-interaction-subscriber~1"]
In addition to performing the actions necessary to subscribe to and receive messages from the transport that they are using, Subscribers in a uProtocol network are *REQUIRED* to invoke the <<subscribe-operation>> and <<unsubscribe-operation>> operations of their local uSubscription service instance in order to indicate which topics they _subscribe_ to or _unsubscribe_ from.

[#subscribe-operation]
=== Subscribe()

// [.specitem,oft-sid="dsn~usubscription-subscribe-protobuf~1",oft-needs="impl,utest"]

> // The uSubscription service `Subscribe()` function *MUST* implement the corresponding protobuf definition in {usubscription_proto_link}.

Reviewer: ```suggestion
```

Author: This one I left in as a marker for this question: do we want to reference the asnycapi defitinition here? 
Because there is nothing fundamentally new/additional contained in the asyncapi spec, it's just a more formalized version of the definitions in the README.

Register the calling client as subscriber to a topic. 

* Returns the status of the requested subscription. 
* Calling this endpoint also registers the subscribing client to receive subscription change notifications when the subscription status changes.

[.specitem,oft-sid="dsn~usubscription-subscribe-method_id~1",oft-needs="impl,utest"]
The uSubscription service `Subscribe()` function *MUST* use the method ID 1.

[.specitem,oft-sid="req~usubscription-subscribe-signature~1",oft-needs="dsn,impl,utest"]
The uSubscription service `Subscribe()` function *MUST* implement the following message definitions:

.Subscribe() function request parameters
[#subscribe-request-data-model]

> [mermaid]

Reviewer: IMHO we should move this class diagram to the top of the file and rename the _Subscribe_ class to _USubscription_ (keeping the interface stereotype) and define the _subscribe_ and _unsubscribe_ operations on the interface ...

Author: Ok, got it.

ifdef::env-github[[source,mermaid]]
----
classDiagram
class Subscribe {
        <<Interface>>
}

class SubscribeRequest {
    <<UMessage>>
    topic : UURI
    expiry : UInt64[0..1]
    sample-period : UInt32[0..1]
}
class SubscribeResponse {
    <<UMessage>>
    topic : UURI
    status : SubscriptionStatus
}

> Subscribe ..> SubscribeRequest : request message

Reviewer: Maybe we can use https://github.com/eclipse-uprotocol/up-spec/blob/main/up-l1/README.adoc#3-utransport as a template for defining the USubscription interface?

Author: I gave it a go - pls have a look at the updated files.

A _Subscriber_ is a uEntity that is interested in a particular _topic_ and wants to receive messages that have been published to that topic by other uEntities. Subscribers typically use the {communication_layer_api_ref} for this purpose.

[.specitem,oft-sid="dsn~usubscription-interaction-subscriber~1"]
In addition to performing the actions necessary to subscribe to and receive messages from the transport that they are using, Subscribers in a uProtocol network are *REQUIRED* to invoke the <<subscribe-operation>> and <<unsubscribe-operation>> operations of their local uSubscription service instance in order to indicate which topics they _subscribe_ to or _unsubscribe_ from.

[#subscribe-operation]
=== Subscribe()

[source]
----

> subscribe(topic: UUri, expiry: UInt64[0..1], sample-period: UInt32[0..1])

Reviewer: ```suggestion
subscribe(topic: UUri, expiry: UInt64 [0..1], sample-period: UInt32 [0..1]) : UUri, SubscriptionStatus
```

Author: ok

----

Clients use this method to subscribe to a topic. 

* Returns the status of the requested subscription. 
* Calling this endpoint also registers the subscribing client to receive subscription change notifications when the subscription status changes.

[.specitem,oft-sid="dsn~usubscription-subscribe-method_id~1",oft-needs="impl,utest"]
The uSubscription service `Subscribe()` function *MUST* use the method ID 1.

[.specitem,oft-sid="req~usubscription-subscribe-signature~1",oft-needs="dsn,impl,utest"]
The uSubscription service `Subscribe()` function *MUST* implement the following message definitions:

---

.Subscribe Request Parameters
[%autowidth]
|===
|Parameter | Type | Mandatory | Description

| topic
| {uuri_ref}
| yes
a|
--
uProtocol URI of the topic to subscribe to.

Also refer to xref:topic-uuris[Topic UURIs] regarding specific restrictions on topic UURIs.
--

|`expiry`
|UInt64
|no
a|
[.specitem,oft-sid="dsn~usubscription-subscribe-past_expiry~1",oft-needs="impl,utest",oft-tags="uSubscription"]
--

> Define when the subscription should expire, in millisecond granularity (timestamp should be set in the future).

Reviewer: IMHO this needs some rewording to be more precise as to what the value should represent. My feeling is that it is supposed to be an absolute point in time (in the future) given as the number of millis since Unix Epoch, right?

Author: done

* If this field is missing, the subscription is assumed to remain valid indefinitely.
* If the timestamp is set in the past when evaluated by uSubscription service, it *MUST NOT* register the subscription and return `UNSUBSCRIBED` as subscription status.
--

|`sample-period`
|UInt32
|no
a|
--
Desired sampling period (in milliseconds) the subscriber wishes to receive
events in, applicable only for remote topics. Device dispatchers (i.e. streamers) use
this attribute to reduce the publication rate of messages sent between devices.
This attribute is might e.g. be used by mobile/cloud clients subscribing to in-vehicle 
topics that in their original scope are published at a high rate.

* If the desired sampling period is lower than the original publisher's publication period, the attribute is ignored.
* If this attribute is missing, the sampling period is chosen by the publisher.
--

|===

.Subscribe Response Parameters
[%autowidth]
|===

> |Name |Type |Description
> 
> |`topic`
> |{uuri_ref}
> a|
> --
> Subscribed topic, as passed to SubscriptionRequest
> --
> 
> |`status`
> |xref:SubscriptionStatus[SubscriptionStatus]
> a|
> --
> Current status of the subscription.
> --
> 
> |===

Reviewer: ```suggestion
[%autowidth]
|===
|Type |Description

|{uuri_ref}
a|
--
Subscribed topic, as passed to SubscriptionRequest
--

|xref:SubscriptionStatus[SubscriptionStatus]
a|
--
Current status of the subscription.
--

|===
```

Author: ok

When uSubscription service receives a reply to a remote `Subscribe()` request, uSubscription service *MUST* set the subscription state for any client-topic combination involving the subscribed remote topic to match the subscription status response of the remote uSubscription service (e.g. `SUBSCRIBED` or `UNSUBSCRIBED`).

[.specitem,oft-sid="req~usubscription-subscribe-unsubscribe-pending~1",oft-needs="impl,utest"]
When a client calls the `Subscribe()` function for a remote topic that is in state `UNSUBSCRIBE_PENDING`, uSubscription service *MUST* initiate the regular remote subscription process, i.e. send a subscription request to the remote uSubscription service and set status to `SUBSCRIBE_PENDING`.

[.specitem,oft-sid="req~usubscription-subscribe-notifications~1",oft-needs="impl,utest"]
When a client calls the `Subscribe()` function, uSubscription service *MUST* generate subscription change notifications reflecting any changes to the subscription state of the subscribed topic.

---

[discrete]
==== Subscribe operation
[mermaid]
ifdef::env-github[[source,mermaid]]
----
sequenceDiagram

actor C as Client
participant T as USubscription Service

> C->>T : subscribe(topic: UUri)

Reviewer: Sequence diagrams usually describe concrete scenarios, the operation's parameters are therefore usually concrete values, e.g.

```suggestion
C->>T : subscribe("up://vehicle/A1B/1/9A10")
```

Author: ok

activate T

> T--)C : topic: UUri, status: SubscriptionStatus

Reviewer: same goes for the return values

```suggestion
T--)C : "up://vehicle/A1B/1/9A10", SubscriptionStatus.subscribed
```

Author: ok

When receiving a `Reset()` call from a source that is not another uSubscription services (i.e. from source URIs where uEntity ID (`ue_id`) does not equal _0x0_), uSubscription service *MUST* return a failure status message with {ucode_link} `PERMISSION_DENIED`.
****

[#common-objects]
== Common Objects

Objects commonly used throughought this document are defined in this section.

[#UURI]
=== UURI

A UURI identifies resources in a uProtocol service mesh, and is defined in link:../../../basics/uri.adoc[the uProtocol URI specification].

[#topic-uuris]
==== Topic UURIs

[.specitem,oft-sid="dsn~usubscription-valid-topic-uuris~1",oft-needs="impl,utest"]
UURIs used to denote topics in uSubscription service API calls *MUST* adhere to the following rules, otherwise the API call will return a failure status message with {ucode_link} `INVALID_ARGUMENT`:

* is not a valid {uuri_ref} or
* contains a _wildcard_ authority or

> * contains a _wildcard_ uEntity ID (`ue_id`) or

Reviewer: Can I subscribe to `up://vehicle/FFFFA1B0/8001`?

> // * contains a _wildcard_ resource ID,

Reviewer: Is this commented out by mistake?

[#subscriber-uuris]
==== Subscriber UURIs

[.specitem,oft-sid="dsn~usubscription-valid-subscriber-uuris~1",oft-needs="impl,utest"]
UURIs used to denote subscribers in uSubscription service API calls *MUST* adhere to the following rules, otherwise the API call will return a failure status message with {ucode_link} `INVALID_ARGUMENT`:

* is not a valid {uuri_ref} or
* contains a _wildcard_ authority or
* contains a _wildcard_ uEntity ID (`ue_id`) or
* contains a _wildcard_ resource ID,

[#subscription-status]
=== SubscriptionStatus

SubscriptionStatus is an enum-style object, describing the status of a subscriber-topic relationship.

.SubscriptionStatus
[cols="1,2,3"]
|===

> | Enum value | Name | Description

Reviewer: FMPOV we do not need the Enum Value column here as it does not serve any purpose

Author: true
